### PR TITLE
[SPARK-2848] Shade Guava in uber-jars.

### PR DESCRIPTION
**Pull request by vanzin**
For further discussion, please check the JIRA entry.

This change moves Guava classes to a different package so that they don't conflict with the user-provided Guava (or the Hadoop-provided one). Since one class (Optional) was exposed through Spark's public API, that class was forked from Guava at the current dependency version (14.0.1) so that it can be kept going forward (until the API is cleaned).

Note this change has a few implications:
- _all_ classes in the final jars will reference the relocated classes. If Hadoop classes are included (i.e. "-Phadoop-provided" is not activated), those will also reference the Guava 14 classes (instead of the Guava 11 classes from the Hadoop classpath).
- if the Guava version in Spark is ever changed, the new Guava will still reference the forked Optional class; this may or may not be a problem, but in the long term it's better to think about removing Optional from the public API.

For the end user, there are two visible implications:
- Guava is not provided as a transitive dependency anymore (since it's "provided" in Spark)
- At runtime, unless they provide their own, they'll either have no Guava or Hadoop's version of Guava (11), depending on how they set up their classpath.

Note that this patch does not change the sbt deliverables; those will still contain guava in its original package, and provide guava as a compile-time dependency. This assumes that maven is the canonical build, and sbt-built artifacts are not (officially) published.

### CODE REVIEW

**Comment by mfawzymkh** [comment]
do we have an ETA to get this pull request merged to master?  The guava shading issue is causing a problem for client libs that has a dependency on swift-service  when spark is compiled with hadoop-2.4

**Comment by srowen** [comment]
@mfawzymkh As you can see this was merged into Spark a long time ago. What are you waiting on?

**Comment by mfawzymkh** [comment]
Sorry , I confused this with some other issue.  Thanks

File: assembly/pom.xml [hunk 1]
```diff
@@ -43,6 +43,12 @@
   </properties>
 
   <dependencies>
+    <!-- Promote Guava to compile scope in this module so it's included while shading. -->
+    <dependency>
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+      <scope>compile</scope>
```
Comment (liuzhongshu) [inline assembly/pom.xml]: why use compile here

**Pull request closed without merging**